kvserver: avoid reusing proposal in tryReproposeWithNewLeaseIndex

tbg opened the pull request:
The previous reproposal mechanism is quite complex. We can side-step a
good amount of the complexity by not re-using proposals. Instead, when
the entry associated to a local proposal gets rejected by the
`LeaseAppliedIndex`, we mint a "new" proposal (identical to the old but
with a new command ID) which is now responsible for notifying the
caller.

This is introduced piecemeal, and the "old way" is preserved for continued
metamorphic testing (and a guaranteed way to be able to fall back, should
we see fallout over the next couple of weeks).

Phasing out the gating var `useReproposalsV2` is tracked separately[^1].

[^1]: https://github.com/cockroachdb/cockroach/issues/105625

Epic: CRDB-25287
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97779)
<!-- Reviewable:end -->

tbg committed:
kvserver: add useReproposalsV2 feature flag

Also, conditionally skip two tests based on it, as these tests will surely break
with any change to the reproposal internals.

Epic: none
Release note: none

tbg committed:
kvserver: introduce (unused) ProposalData.v2SeenDuringApplication

This will be set from code that is yet to be written to address
undefined/possibly data-racy access to ProposalData. We haven't seen issues
with this in practice, but since ProposalData may exist outside of the
proposals map during log application and it will be accessed and mutated
without replicaMu while *also* being reinserted into the proposals map, I don't
see any reason why data races wouldn't be possible. We're probably just getting
lucky today.

This paves the road to not having to rely on luck in the future.

tbg committed:
kvserver: add TODO on signalProposalResult

tbg committed:
kvserver: set response error on rejected proposal

tbg committed:
kvserver: split tryReproposeWithNewLeaseIndex

The first part is annoying because it `return nil` in cases where V2 would just
want to error out. To avoid this branching between v1 and v2 to become too
spaghetti, split the "bad" from the "good".

tbg committed:
kvserver: remove a misleading TODO

I think this TODO was too optimistic. Yes, a Replica will assign a unique
LeaseAppliedIndex to each proposal it produces. However, this is not true across
Replicas, so we need to additionally throw in the proposer's ReplicaID.
Additionally, the LeaseAppliedIndex counter is ephemeral, so a Replica may still
reuse the IDs and we need yet another component (like a "process instance
counter").

We may as well stick to the 8 bytes of entropy we have today.

tbg committed:
kvserver: add retrieveLocalProposalsV2

retrieveLocalProposalsV2 is used with useReproposalsV2, replacing a call to
retrieveLocalProposals. The V2 implementation is simpler because a log entry
that comes up for a local proposal can always consume that proposal from the map
because V2 never mutates the MaxLeaseIndex for the same proposal. In contrast,
with V1, we can only remove the proposal from the map once we have found a log
entry that had a matching MaxLeaseIndex. This lead to the complexity of having
multiple entries associated to the same proposal during application.

tbg committed:
kvserver: add tryReproposeWithNewLeaseIndexV2

With this, we're in a position to actually enable `useReproposalsV2`: we now
repropose as a "new" proposal.

The process of "forking" a proposal into a new proposal is still too manual and
there is little focused testing coverage. However, adding such coverage likely
incurs many refactors. So instead here the approach is to make minimal changes
and rely on randomized coverage via
`TestProposalsWithInjectedLeaseIndexAndReproposalError` to weed out "obvious"
issues and have some confidence that the changes "work" at least most of the
time.

At that point we can commit to moving forward with V2 and start simplifying the
code, which also allows adding more targeted testing (though this will still be
challenging as some of the invariants we'd like to test are quite subtle and
spread out).

tbg committed:
kvserver: add some comments inside ProposalData

tbg committed:
kvserver: improve comments on replication trace spans

It was pretty confusing to me at first. It still
is somewhat confusing, but at least now it's
also somewhat documented.

erikgrinaker commented on pkg/kv/kvserver/replica_application_decoder.go:
```diff
@@ -73,10 +73,72 @@ func (d *replicaDecoder) decode(ctx context.Context, ents []raftpb.Entry) error
+			// (propBuf.{Insert,ReinsertLocked} ignores proposals that have
+			// v2SeenDuringApplicationSet to make this true).
+			if cmd.proposal.v2SeenDuringApplication {
+				log.Fatalf(d.r.AnnotateCtx(context.Background()), "ProposalData seen twice during application: %+v", cmd.proposal)
```
Can we make this not be fatal in release builds? Consider `logcrash.ReportOrPanic`.

erikgrinaker commented on pkg/kv/kvserver/replica_application_decoder.go:
```diff
@@ -73,10 +73,72 @@ func (d *replicaDecoder) decode(ctx context.Context, ents []raftpb.Entry) error
+			delete(d.r.mu.proposals, cmd.ID)
+			if d.r.mu.proposalQuota != nil {
+				alloc = cmd.proposal.quotaAlloc
+				cmd.proposal.quotaAlloc = nil
```
nit: this isn't strictly necessary any longer, right, since we can no longer apply twice? Although I suppose it would be if we made the above assertion non-fatal.

tbg replied:
You mean we could leave the quota allocation on the proposal? We might get away with it but it doesn't seem clean.
The quota allocation moves to the new proposal in `tryReproposeWithNewIndex`, which semantically seems right (though de facto it is always nil at that point because of this code here). And ignoring reproposals, it still seems better to "unhook" it here, since the `quotaAlloc` moves to the `quotaReleaseQueue`.

erikgrinaker commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -225,6 +280,160 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
+	// or release any latches.
+
+	origP := origCmd.proposal
+	if origP.v2Reproposed {
```
Isn't this assertion identical to the one above?

erikgrinaker commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -225,6 +280,160 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
+		// has to be local to enter this branch, there is a bug - we should not see
+		// it as local twice in a row.
+		//
+		// TODO(tbg): it's likely that the `v2Reproposed` field isn't adding any
```
I was wondering the same thing. Can we put this TODO on the field itself?

tbg replied:
I removed the field.

erikgrinaker commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -225,6 +280,160 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
+		}
+	}()
+
+	newCommand := kvserverpb.RaftCommand{
```
We should have a comment on `RaftCommand` pointing out that any changes to it must be reflected here. Alternatively (and better), use a method like we do during Raft application where we zero out the fields we've processed and then assert that the remaining struct is zero-values.

I see you mentioned a unit test below. That would also work.

tbg replied:
Going with the unit test for now. The semantics of what needs to be set when, etc, are tricky. What matters most is that we are forced to think about it when adding a new field.

erikgrinaker commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -225,6 +280,160 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
+		sp:              origP.sp, // NB: special handling below
+		idKey:           raftlog.MakeCmdIDKey(),
+		proposedAtTicks: ticks,
+		createdAtTicks:  ticks,
```
We can get rid of one of these `AtTicks` fields once this migration is fully complete, right?

tbg replied:
Why? The purpose of these fields is to let us determine the age of a proposal (to determine whether to repropose it in `refreshProposalsLocked`: https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvserver/replica_raft.go#L1380

erikgrinaker replied:
Hm, right. Shouldn't we carry over `createdAtTicks` from the original proposal in that case? Or is the idea that we got an actual error result on the previous proposal, so this is a new proposal for all intents and purposes?

Since we use this for observability, and users presumably only care about how long the original request takes, it seems to make sense to retain the original `createdAtTicks`.

tbg replied:
Small correction, `createdAtTicks` is used for the circuit breaker, but if we get proposals through raft and pick new lease applied indexes, I don't think that should end up circuit-breaking the replica (though it sure would be odd if it happened for 1000s of round-trips for a single user command...).
The refresh mechanism only uses `proposedAtTicks`.

> Since we use this for observability, and users presumably only care about how long the original request takes, it seems to make sense to retain the original createdAtTicks.

See above, I am leaning the other way for this PR. It can go either way, honestly, but I'd rather introduce "proposal-spanning" properties deliberately.

Where do we use this for observability? Ticks aren't a great unit of measurement for user-facing purposes anyway since it's so discrete. We could change `proposedAtTicks` into a `time.Time`, then keep it across reproposals, and then when the proposal is signaled report a metric (user-visible replication latency). For our internal latency metric, we'd want to measure each individual round-trip through the replication layer, though, which would mean adding another timestamp that _is_ reset on reproposals. I think we'd best do things like that separately though, in the context of https://github.com/cockroachdb/cockroach/issues/83262.

erikgrinaker replied:
I thought we used it to detect slow proposals, but you're right that it's only used for the circuit breaker. This seems fine then.

erikgrinaker reviewed (approved):
Really appreciate the simplification, and the well-structured PR! I'm partial to adding some tests/assertions ensuring that new fields added to the command/proposal structs will have to be processed during reproposals, but we can do that later if you pinky-swear that it'll be soon.

Also, this area is really subtle, and I'm not deeply familiar with all the nooks and crannies so I haven't been able to review this exhaustively, but it LGTM. I'm trusting that the added test coverage does a better job at picking up issues anyway.

tbg committed:
kvserver: add TODO question on flow control token logic

tbg committed:
kvserver: add tripwire test for adding fields to ProposalData and RaftCommand

When a field is added to these, two places need to be updated - the main
proposal path and tryReproposeWithNewLeaseIndex{,v2}. This commit adds a
test that will explicitly need to be updated when such a field is added.

Epic: none
Release note: None

erikgrinaker reviewed (approved):
LGTM, thanks!

tbg commented:
> ./dev test --stress ./pkg/kv/kvnemesis/ --filter ReproposalChaos --stress-args '-timeout 2m' --cpus 4
> 4283 runs so far, 0 failures, over 1h35m40s

I guess I'm doing this! TFTR Erik.

bors r=erikgrinaker
